zio-http/src/main/scala/zio/http/model/Header.scala

```diff
@@ -0,0 +1,4238 @@
+package zio.http.model
+
+import java.net.URI
+import java.nio.charset.StandardCharsets
+import java.time.format.DateTimeFormatter
+import java.time.{ZoneOffset, ZonedDateTime}
+import java.util.Base64
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.util.control.NonFatal
+import scala.util.matching.Regex
+import scala.util.{Either, Failure, Success, Try}
+
+import zio._
+
+import zio.http._
+import zio.http.model.Header.Authorization.AuthScheme
+import zio.http.model.Header.HeaderType.Aux
+
+sealed trait Header {
+  def headerName: CharSequence
+  def renderedValue: CharSequence
+}
+
+object Header {
+  sealed trait HeaderType {
+    type HeaderValue
+
+    def name: CharSequence
+
+    def parse(value: String): Either[String, HeaderValue]
+
+    def render(value: HeaderValue): String
+  }
+
+  object HeaderType {
+    type Aux[HV] = HeaderType { type HeaderValue = HV }
+  }
+
+  sealed trait HeaderValue[H <: Header] { self: H =>
```

Comment:
Do we need a distinction between `Header` and `HeaderValue`?

Conceptually, I view a header as a combination of two things: the type (e.g. `Content-Type`), and the value (e.g. `application/json`). 

If we can move from `Header` + `HeaderValue` to just `Header`, it will be a simpler API.

Reply:
I wanted two things with it:

- Have a simple untyped `Header` interface with no type parameters, to be used to get the name and value as strings when for example converting to netty data types, etc. 
- The other to have the actual type as a type parameter, so type inference works properly (without it, ```header.headerType.render(header)``` was not type checking for example)

Reply:
It could be just `Header[H]` and used as `Header[_]` in places where we don't care. Is that better?